### PR TITLE
OPENEUROPA-1292: OE Webtools Analytics refactoring

### PR DESCRIPTION
**Pull request by pfrenssen**
### Description

Refactoring.

### Change log

- Added: No
- Changed: Yes
- Deprecated: No
- Removed: No
- Fixed: No
- Security: No

### Commands

```sh
$ sudo make me a sandwich
```

### CODE REVIEW

**Comment by pfrenssen** [comment]
The test shows that the result returned by the rules based analytics doesn't vary by the page URI. This is a bug; since the rules are URI based the result should be cached individually for each URI.

This is currently only working because `AnalyticsEventSubscriber::onSetSiteDefaults()` also varies by URI, but we shouldn't rely on event subscribers from different modules to take care of our responsibilities.

**Comment by pfrenssen** [comment]
I expanded the test, I now exhibits a failure that is occurring if the request stack is updated after the service is instantiated. The problem is that current request path is being cached on instantiation but since it persists on the container it shouldn't do this. It should instead retrieve the request on demand.

**Comment by pfrenssen** [comment]
The failure is due to a bug that causes the current path to be overwritten with the path in the site's default language as soon as a rule is processed that has the option to match on the default language enabled. All subsequent rules are being processed as if this option is set to `TRUE`.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+  Ref.
+  [OPENEUROPA-1633](https://webgate.ec.europa.eu/CITnet/jira/browse/OPENEUROPA-1633)
+* The option to match multilingual content on the path alias for the default
+  language only works if a translation exists in the default content of the
+  website.
+
+  Ref.
+  [OPENEUROPA-1637](https://webgate.ec.europa.eu/CITnet/jira/browse/OPENEUROPA-1637)
+* There currently is no support for multilingual content based on system paths.
```
Comment (brummbar) [inline modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/README.md]: I cannot reproduce this issue. `\Drupal\Core\Path\AliasManagerInterface::getAliasByPath()` documentation says that `An alias that represents the path, or path if no alias was found.`.
I tried with a rule for `/admin/people` and it worked for the unaliased default language and for the aliased translated one. Same for a node created in non-default language.

File: modules/oe_webtools_analytics/tests/src/Functional/ConfigurationTest.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    'config',
+    'system',
+    'oe_webtools',
+    'oe_webtools_analytics',
+  ];
+
+  /**
+   * Tests if the configuration for the Webtools library is present on the page.
+   *
+   * @throws \Behat\Mink\Exception\ExpectationException
```
Comment (brummbar) [inline modules/oe_webtools_analytics/tests/src/Functional/ConfigurationTest.php]: This is not needed.

File: modules/oe_webtools_analytics/tests/src/Functional/ConfigurationTest.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+   *   Thrown when an expected JSON configuration object is not present in the
+   *   page.
+   */
+  public function testLibraryLoading(): void {
+    $config = \Drupal::configFactory()
+      ->getEditable(AnalyticsEventInterface::CONFIG_NAME)
+      ->set("siteID", "123")
```
Comment (brummbar) [inline modules/oe_webtools_analytics/tests/src/Functional/ConfigurationTest.php]: We can take the chance to replace these double quotes with single ones.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+ * Service that matches routes to rules.
+ */
+class RuleMatcher implements RuleMatcherInterface {
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
```
Comment (brummbar) [inline modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: Nit: can we sort alphabetically the properties? 😛

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    $cache->data = ['rule' => $rule_id] + ($cache->data ?? []);
+
+    // We return results based on rule entities. This means that if a rule is
+    // added or deleted, or if any of the existing rules change, the cached
+    // results should be invalidated.
+    $cache->tags = Cache::mergeTags($cache->tags ?? [], $this->getWebtoolsAnalyticsRuleListCacheTags());
+
+    // Add the cache tags of the default site configuration if the rule depends
+    // on the default language of the site.
+    $default_language_cache_tags = $matching_rule instanceof WebtoolsAnalyticsRuleInterface && $matching_rule->matchOnSiteDefaultLanguage() ? $site_configuration->getCacheTags() : [];
```
Comment (brummbar) [inline modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: Let's use a more clear-to-understand `if()` in here for code readability.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/EventSubscriber/WebtoolsAnalyticsRulesEventSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  /**
+   * Returns the entity type definition for the Webtools Analytics Rule entity.
+   *
+   * @return \Drupal\Core\Entity\EntityTypeInterface
+   *   The entity type definition.
+   */
+  protected function getWebtoolsAnalyticsRuleDefinition(): EntityTypeInterface {
+    try {
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/EventSubscriber/WebtoolsAnalyticsRulesEventSubscriber.php]: Welllll.... I think this is not a questionable practice at all but a very useful feature that trickled down from Java to PHP and is leveraged effectively by the exception classes in Drupal core.I think this feature is sadly overused and it's very positive that this is being pointed out by PHPStorm. If this would not be useful then core could just throw unchecked `RuntimeExceptions` for everything. But they are not, except when the error is a runtime condition which can not be dealt with higher up the stack.

The whole idea behind checked exceptions is that they _have to be dealt with_. Stricter languages like Java won't even compile if there are any checked exceptions allowed to bubble up to the top of the call stack. 

For this particular piece of code this is the right spot for handling these particular exceptions. We are calling into a dynamic system that will throw checked exceptions in case invalid input is given. But we know for a fact that we are passing valid data since we define the entity type ourselves. We can safely convert the exception so the calling code does not need to be concerned about handling these exceptions. After all why would an event dispatcher even be concerned with catching checked exceptions for the entity type manager?

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/WebtoolsAnalyticsRuleListBuilder.php [hunk 7]
```diff
@@ -27,6 +28,9 @@ public function buildHeader(): array {
    * {@inheritdoc}
    */
   public function buildRow(EntityInterface $entity): array {
+    if (!$entity instanceof WebtoolsAnalyticsRuleInterface) {
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/WebtoolsAnalyticsRuleListBuilder.php]: We cannot be sure of that, according to the interface any entity type can be passed in, but we can only deal with rule entities. If I didn't add this check here I would not be able to call custom methods like `::getSection()` and `::getRegex()` but would have to retrieve data from the EntityFieldManager, making the below code a lot more complex than needed.

File: modules/oe_webtools_analytics/tests/src/Functional/ConfigurationTest.php (hunk 2 above)
Comment (drupol) [reply modules/oe_webtools_analytics/tests/src/Functional/ConfigurationTest.php]: I have a question regarding your comment @brummbar.

I guess @pfrenssen added this because the method throws an exception at a certain point and probably PHPStorm suggested to add it.
Is it a so bad thing to add it in the phpdoc? Why would we remove this?

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/EventSubscriber/WebtoolsAnalyticsRulesEventSubscriber.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/EventSubscriber/WebtoolsAnalyticsRulesEventSubscriber.php]: The value is that we are responsibly handling checked exceptions. The `catch (PluginNotFoundException $e)` is the whole point of this code. The conversion into a `\RuntimeException` is just a small bonus. But since we are calling into code that throws checked exceptions it is our responsibility to catch all of them.

I tried to explain this in my previous comment but it doesn't seem to be clear. I'll elaborate a bit.

**Checked vs unchecked**

Exceptions are thrown when an error occurs in the code which is outside of the scope of what that particular method is designed to deal with. There are two types of problems for which exceptions are thrown: either the problem is something that the calling code can deal with (e.g. trying to load a user with a name that doesn't exist), or it is something completely out of control of the codebase (e.g. the disk is full, or the database server is down). The first type is called a "checked exception" (because it should be "checked" (caught) by the calling code) and the second is an "unchecked exception" since it cannot be handled and should bubble up to the root of the call stack or the generic error handler.

PHP doesn't have the concept of checked exceptions built in to the language, but by convention all exceptions except `\RuntimeException` are considered to be checked exceptions. This is in fact identical to how it is handled in Java, except that PHP will not throw errors at compile time. Luckily we can detect them with static analysis. PHPStorm (being based on IntelliJ) has very good support for it.

Drupal core is (for the most part) correct in their use of emitting the right exception type. For example the entity type manager (as any other plugin manager) will throw a checked `PluginNotFoundException` if it is asked to act on an entity type that it doesn't know about. The idea is that is should be handled by the calling code: they can provide fallbacks (by implementing `FallbackPluginManagerInterface`), ignore it if it is an optional plugin, or show relevant information to the user (for example an error message, or a form that allows to create or buy the plugin).

Here is an illustration of this principle in action, from `ContentModerationConfigureEntityTypesForm`:

```
  public function buildForm(array $form, FormStateInterface $form_state, WorkflowInterface $workflow = NULL, $entity_type_id = NULL) {
    try {
      $this->entityType = $this->entityTypeManager->getDefinition($entity_type_id);
    }
    catch (PluginNotFoundException $e) {
      throw new NotFoundHttpException();
    }
    // ...
  }
```

This form gets an entity type ID as user input, passes it on to the entity type manager, which throws a checked exception if the plugin type doesn't exist. This checked exception is caught by the calling code who knows how to deal with it, and converted into a `NotFoundHttpException` which is (you guessed it) an unchecked exception which extends `\RuntimeException`.

Checked exceptions that are not handled are indicative of programming errors. They should either be checked (caught), or documented on all the public methods that possibly throw them, so that it is clear to the calling code that _they_ are responsible for catching them. In this particular example the calling code is the form builder which has no knowledge of the entity type manager, so it cannot handle it. That means this code _must_ catch it. It does and decides to let this bubble up as an HTTP error to the end user.

**Handling errors that "will never be thrown"**

Here is some information written by our colleague Andreas on the matter:
https://stackoverflow.com/questions/35586882/php-or-other-strategy-to-deal-with-exceptions-that-cannot-occur

So what is the point of converting this `PluginNotFoundException` into a `RuntimeException` if the `PluginNotFoundException` can never be thrown? I agree that on the surface it can seem pointless and not bringing value. But as explained the actual value is in the handling (catching) of the checked exception, not in the rethrowing.

So why do I do this anyway? Let's have a look at typical ways to handle "impossible" checked exceptions:

*The empty catch block*

```
try {
  $definition = $this->entityTypeManager->getDefinition('webtools_analytics_rule');
}
catch (PluginNotFoundException $e) {
  // This exception not occur.
}
```

The simplest way to deal with checked exceptions is to have an empty `catch {}` block. This is very commonly used. In many use cases this is fine, since the exception is not supposed to happen, right?

But there is a catch in highly complex systems like Drupal (pun intended). It is not unheard of that this "impossible" exception still happens in some rare edge case. For example the entity type manager might rely on data cached in Redis. At some point the Redis server gets out of sync due to a bug unrelated to our code, and suddenly the entity type manager thinks that our entity type doesn't exist and throws the "impossible" exception. If we ignore it completely then we have no trace of the runtime problem that occurred.

*Logging*

```
try {
  $definition = $this->entityTypeManager->getDefinition('webtools_analytics_rule');
}
catch (PluginNotFoundException $e) {
    $this->logger->warning('The webtools_analytics_rule entity type does not exist.');
}
```

One common way of still keeping track of "impossible" exceptions that are being thrown is to log them. This has some overhead of having to inject the logger service.

*Rethrowing as an unchecked exception*

```
try {
  $definition = $this->entityTypeManager->getDefinition('webtools_analytics_rule');
}
catch (PluginNotFoundException $e) {
  throw new \RuntimeException('The webtools_analytics_rule entity type does not exist.', 0, $e);
}
```

The third possibility is to simply rethrow the "impossible" exception as an unchecked `\RuntimeException`. This clearly communicates to the calling code that if this goes wrong it is because of a runtime problem, and we still get the advantages of Drupal's error handling that will log the error and show a message to the user.

So basically converting the checked exception into an unchecked exception is a convenient and robust way to handle this "exceptional" situation.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+    $current_path = $this->getCurrentPath();
+    $cache = $this->cache->get($current_path) ?: new \stdClass();
+
+    // Check if the cache data for the section has been set, taking into account
+    // that its value might be NULL.
+    if (empty($cache->data) || !array_key_exists('section', $cache->data)) {
+      $rule = $this->getMatchingRule($current_path);
+
+      // Reload the cache object, it is updated when the rule was matched.
+      $cache = $this->cache->get($current_path) ?: new \stdClass();
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: I had to add it because the tests coverage pointed out that the cache items were not correctly invalidated if a rule is matching on the site default language and the default language is being changed. It needs to have the cache tags added of the default site configuration, but only if the rule matches on the site default language.

The cache entry is not per section or per rule, but per path, so this is fine. A path entry has two data points: the section to which the path belongs and the rule that is responsible for setting it.

I'll have a look to refactor this so that the weird cache juggling is not being done. It will be cleaner and simpler if we just populate both on a cache miss.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+    $matching_rule = NULL;
+
+    foreach ($this->loadWebtoolsAnalyticsRules() as $rule) {
+      if (preg_match($rule->getRegex(), $rule->matchOnSiteDefaultLanguage() ? $default_language_alias_path : $path) === 1) {
+        $matching_rule = $rule;
+        break;
+      }
+    }
+
+    $rule_id = $matching_rule instanceof WebtoolsAnalyticsRuleInterface ? $matching_rule->id() : NULL;
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: OK I will split the ternaries over more lines.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+    $cache->data = ['rule' => $rule_id] + ($cache->data ?? []);
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: Because there could be an existing cache entry for 'section' but not for 'rule', or it could be `NULL`.

I'll refactor this so this cache won't be half populated any more.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+   * Returns the Webtools Analytics Rule entities with the given entity IDs.
+   *
+   * @param string[]|null $ids
+   *   Optional array of entity IDs to return. If omitted all entities will be
+   *   returned.
+   *
+   * @return \Drupal\oe_webtools_analytics_rules\Entity\WebtoolsAnalyticsRule[]
+   *   The entities.
+   */
+  protected function loadWebtoolsAnalyticsRules(array $ids = NULL): array {
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: Absolutely :)

File: modules/oe_webtools_analytics/tests/src/Functional/ConfigurationTest.php (hunk 2 above)
Comment (pfrenssen) [reply modules/oe_webtools_analytics/tests/src/Functional/ConfigurationTest.php]: Hmm yeah this is actually quite pointless in tests to say which exception causes the tests to fail. This is left over from an earlier iteration where code that throws this exception was in a separate protected method. It looks like this code doesn't  even made it to the final version. I'll remove it.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php (hunk 4 above)
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: Oh yiss :smile_cat: I'll even inject them alphabetically :laughing:

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/README.md (hunk 1 above)
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/README.md]: Interesting, I also see that there are no test cases referencing this problem. Possibly it has been accidentally solved as part of this issue. I will double check and see if we have coverage for this use case.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+      // Reload the cache object, it is updated when the rule was matched.
+      $cache = $this->cache->get($current_path) ?: new \stdClass();
+
+      $section = $rule instanceof WebtoolsAnalyticsRuleInterface ? $rule->getSection() : NULL;
+      $cache->data = ['section' => $section] + ($cache->data ?? []);
+
+      // We return results based on rule entities. This means that if a rule is
+      // added or deleted, or if any of the existing rules change, the cached
+      // results should be invalidated.
+      $cache->tags = Cache::mergeTags($cache->tags ?? [], $this->getWebtoolsAnalyticsRuleListCacheTags());
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: OK I need to address this when I refactor the RuleMatcher.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+    $this->currentPath = $currentPath;
+    $this->aliasManager = $aliasManager;
+    $this->cache = $cache;
+    $this->config = $config;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getMatchingSection(string $path = NULL): ?string {
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: Woopsie! Fixed :)

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: I shortened:

* `::getWebtoolsAnalyticsRuleListCacheTags()` ☞ `::getListCacheTags()`
* `::loadWebtoolsAnalyticsRule()` ☞ `::loadRule()`
* `::loadWebtoolsAnalyticsRules()` ☞ `::loadRules()`
* `::getWebtoolsAnalyticsRuleDefinition()` ☞ `::getRuleDefinition()`
* `::getWebtoolsAnalyticsRuleEntityStorage()` ☞ `::getRuleEntityStorage()`

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/README.md (hunk 1 above)
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/README.md]: OK I managed to get it working too. Strangely enough this can be made to work when the option to "match on the default language alias" is *enabled*.  It will then not match on the alias but on the system path :confused:. My test case had this option *disabled* since we're not trying to match on the alias but on the system path. The ticket to fix it proposes to add a third option "Match on system path".

I will update the test case and then the ticket can be closed. I wonder how to explain this case to the end user in the documentation though.

Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/README.md]: OK never mind, it still breaks then after an English language alias is added for the same path. So the bug is still there: it's not possible to reliably create a site section based on the system path - either it doesn't work when there is an alias in the default language, or it doesn't work when there is _no_ alias :confounded:. I'll update the ticket and the README.

Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/README.md]: I have updated the Jira ticket and the information in the "Known issues" section to better explain the problem.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: Fixed.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: Washed the yaks.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: Fixy fixy

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php (hunk 5 above)
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: It hath been understandified.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php]: Splitified.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/tests/src/Kernel/AnalyticsRulesSubscriberTest.php [hunk 14]
```diff
@@ -0,0 +1,478 @@
...
+            // the default language alias.
+            '/taxonomy/term/344' => 'overview of antarctican news (current path)',
+            '/news/antarctica' => 'overview of antarctican news (default language alias)',
+            '/es/nuevas/antartida' => 'overview of antarctican news (default language alias)',
+            // This doesn't match on the system path but on the default language
+            // alias since this rule has a higher priority.
+            '/articles_page' => 'overview of articles (current path)',
+            // There is no rule to define the default language alias when the
+            // site language is set to Spanish, but it should still be possible
+            // to match this using a rule that resolves the system path.
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/tests/src/Kernel/AnalyticsRulesSubscriberTest.php]: My idea is that since this is an open source library it is always possible that the community can develop this feature even if the PO never prioritizes this.

I think being able to resolve system paths is very important. Currently we only support aliases, but aliases can change at will, invalidating any rules that were written. Using system paths and routes is much more resilient.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/tests/src/Kernel/AnalyticsRulesSubscriberTest.php [hunk 15]
```diff
@@ -0,0 +1,478 @@
...
+            //   system path rule. This will require a dependency on the
+            //   Pathauto module.
+            // @see https://webgate.ec.europa.eu/CITnet/jira/browse/OPENEUROPA-1637
+            '/articles_page' => 'overview of articles (current path)',
+            // This is expected to match on the overview of articles in the
+            // default language alias but this is currently not working because
+            // the alias in English doesn't exist.
+            // @todo Update this once OPENEUROPA-1637 is fixed.
+            // @see https://webgate.ec.europa.eu/CITnet/jira/browse/OPENEUROPA-1637
+            '/es/articulos' => '',
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/tests/src/Kernel/AnalyticsRulesSubscriberTest.php]: Great find, I added this rule which matches, and now the bug is correctly proven:

```
'7_articles_overview_default_language_alias' => [
  'section' => 'overview of articles (default language alias)',
  'regex' => '|^/articles_page/?$|',
  'match_on_site_default_language' => TRUE,
],
```

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/tests/src/Kernel/AnalyticsRulesSubscriberTest.php [hunk 16]
```diff
@@ -0,0 +1,478 @@
...
+            // This currently matches on the system path instead of on the
+            // default language alias since there is no alias defined in English
+            // (which is the default language).
+            // @todo Update this one OPENEUROPA-1637 is fixed.
+            //   Once this is fixed this is expected to match on the default
+            //   language alias since this rule has a higher priority than the
+            //   system path rule. This will require a dependency on the
+            //   Pathauto module.
+            // @see https://webgate.ec.europa.eu/CITnet/jira/browse/OPENEUROPA-1637
+            '/articles_page' => 'overview of articles (current path)',
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/tests/src/Kernel/AnalyticsRulesSubscriberTest.php]: There was a bug in the code that I found and tried to prove with this test but now it's been too long and I don't remember it any more. I will remove this test, and hope that the bug will resurface at some point in the future.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/tests/src/Kernel/AnalyticsRulesSubscriberTest.php [hunk 17]
```diff
@@ -0,0 +1,478 @@
...
+    foreach ($expected_sections_by_language as $language => $expected_sections) {
+      $this->config('system.site')->set('default_langcode', $language)->save();
+
+      // Check that the expected sections are returned for the given test paths.
+      foreach ($expected_sections as $path => $expected_section) {
+        // Start with a clean event for each test case.
+        $this->initializeEvent();
+
+        // Set the current path to the one being tested.
+        $this->setCurrentPath($path);
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/tests/src/Kernel/AnalyticsRulesSubscriberTest.php]: Good point, implemented it like this now.

File: modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/tests/src/Kernel/AnalyticsRulesSubscriberTest.php [hunk 18]
```diff
@@ -0,0 +1,478 @@
...
+   * An array of test path aliases in different languages, keyed by system path.
+   */
+  const PATH_ALIASES = [
+    '/news_overview_page' => [
+      'en' => '/news',
+      'es' => '/nuevas',
+    ],
+    '/taxonomy/term/344' => [
+      'en' => '/news/antarctica',
+      'es' => '/es/nuevas/antartida',
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/tests/src/Kernel/AnalyticsRulesSubscriberTest.php]: Ah yes, don't understand how I missed this :-/
